syft/messaging/plan.py

```diff
@@ -190,59 +311,53 @@ def _recv_msg(self, bin_message: bin):
 
-        The plan must be built with some input data, here `args`. When they
-        are provided, they are sent to the plan worker, which executes its
-        blueprint: each command of the blueprint is catched by _recv_msg
-        and is used to fill the plan.
+        The build operation is done "on" the plan (which can be seen like a
+        worker), by running the forward function. The plan will therefore
+        execute each command of the function, will cache it using _recv_msg
+        and will use it to fill the plan.
+        To do so, all the arguments provided and the state elements should be
+        moved to the plan, using send().
 
         Args:
             args: Input data.
         """
-        self._prepare_for_running_local_method()
-        self._build(list(args))
-        self._after_running_local_method()
+        args = list(args)
 
-    def _build(self, args: List):
-        if self.is_method:
-            args = tuple([self._self] + args)
-
-        # The ids of args of the first call, which should be updated when
-        # the function is called with new args
+        # Move the arguments of the first call to the plan and store their ids
+        # as they will be included in the readable_plan: it should be updated
+        # when the function is called with new args and that's why we keep the
+        # refs self.arg_ids
         self.arg_ids = list()
         local_args = list()
         for arg in args:
-            # Send only tensors (in particular don't send the "self" for methods)
-            # in the case of a method.
-            if isinstance(arg, FrameworkTensor):
-                self.owner.register_obj(arg)
-                arg = arg.send(self)
-                arg.child.garbage_collect_data = False
-                self.arg_ids.append(arg.id_at_location)
+            arg = arg.send(self)
+            self.arg_ids.append(arg.id_at_location)
             local_args.append(arg)
 
-        res_ptr = self.blueprint(*local_args)
-
+        # Same for the state element: we send them to the plan and keep reference
+        # to the ids
+        copied_state = self.state.copy()
+        self.state.send(location=self)
+        self.state_ids = self.state.get_id_at_location()
+
+        # We usually have include_state==True for functions converted to plan
+        # using @func2plan and we need therefore to add the state manually
+        if self.include_state:
+            res_ptr = self.forward(*local_args, self.state)
+        else:
+            res_ptr = self.forward(*local_args)
         res_ptr.child.garbage_collect_data = False
 
-        worker = self.find_location(args)
+        # We put back the original state. There is no need to keep a reference
+        # To the element sent to the plan.
+        self.state.set_(copied_state)
 
+        # The readable plan is now built, we hide the fact that it was run on
+        # the plan and on the owner by replacing the workers ids
```

Comment:
the plan and on the owner by replacing the workers ids -> the plan and not on the owner by replacing the workers ids